### PR TITLE
Rewritten __fish_complete_proc.fish for better process completion

### PR DESCRIPTION
**Pull request by pemrmaeoyve**
When I used completions on killall, I noticed that it was displaying kernel Kworker processes and processes that didn't exist, so I decided to fix that.

Before:
screenshot from 2016-05-11 08-57-09
After:
screenshot from 2016-05-11 08-56-14

### CODE REVIEW

**Comment by faho** [comment]
I'm not convinced.

- We want to _reduce_ our reliance on awk and sed and such, preferring our own `string` utility

- This function isn't just used for killall but also for `ps`, where these things might actually be of some use

- `ps eu` for me returns _way_ too few processes, which you can also see in your screenshots - aren't "colord" or "mission-control-5" killable by you?

In conclusion, I don't think this approach works. Sorry!

**Comment by pemrmaeoyve** [comment]
I tried to implement it this way, since zsh does the same thing. 
screenshot from 2016-05-11 14-14-39

1. I didn't know that, and apparently my fish installation doesn't include `string`
2. I could adapt these changes to killall.fish, would that be alright?
3. Those are the processes owned by $USER, but I suppose `ps au` could work too.

**Comment by faho** [comment]
>I didn't know that, and apparently my fish installation doesn't include `string`

It's new in 2.3.0 (which should be released any minute now, we're on the second beta).

>I could adapt these changes to killall.fish, would that be alright?

That would be acceptable.

>Those are the processes owned by $USER, but I suppose ps au could work too.

That doesn't seem to work any better - you absolutely need "x" (or something like it, linux `ps` is this weird frankensteinian monster of disparate implementations), since it will 'Lift the BSD-style "must have a tty" restriction'. Without it, I wouldn't get e.g. my currently running firefox. Which is something I might like to kill all of.

And if you add "x", you'll also get the kernel workers again, so you'd need to filter them out anyway. At that point you could just adjust the current sed call (and I'll convert it to `string` later).

**Comment by pemrmaeoyve** [comment]
Okay, I've adjusted the current sed calls and adapted my changes to killall.fish. 
If the user is root, killall uses __fish_complete_proc and displays all processes, else it just display users' processes.

File: share/functions/__fish_complete_proc.fish [hunk 1]
```diff
@@ -17,16 +17,21 @@ function __fish_complete_proc --description 'Complete by list of running process
...
 		
 		# Retain the last path component only
-		set sed_cmds $sed_cmds 's|.*/||g'
+		set sed_cmds $sed_cmds 's/.*\///'
 		
-		# Strip off square brackets. Cute, huh?
-		set sed_cmds $sed_cmds 's/[][]//g'
+		# Strip off  brackets. Cute, huh?
+		set sed_cmds $sed_cmds '/^\[/d'
+        set sed_cmds $sed_cmds '/\]$/d'
```
Comment (krader1961) [inline share/functions/__fish_complete_proc.fish]: Odd change in indentation here and below. I wouldn't bother fixing it because this module doesn't conform to the style guide. Too, I'm going to blanket restyle all the fish scripts to conform to the style guide in the near future. This is just a FYI comment to the broader community that we really need to be more vigilant about changes that aren't consistent with the style guide.

File: share/completions/killall.fish [hunk 2]
```diff
@@ -22,7 +22,11 @@ if test (uname) != 'SunOS'
 		end
 	end
 
-	complete -c killall -xa '(__fish_complete_proc)'
+    if test $USER = root
```
Comment (krader1961) [inline share/completions/killall.fish]: That var should always be set but it's better to be safe and put double-quotes around it. This is one place where the usual fish semantics aren't in effect in order to be POSIX compliant. If `$USER` isn't set a "missing argument" error will be thrown.

File: share/completions/killall.fish [hunk 3]
```diff
@@ -22,7 +22,11 @@ if test (uname) != 'SunOS'
...
+        complete -c killall -xa '(__fish_complete_proc)'
+    else
+       complete -c killall -xa "(ps ua | awk '{print \$11}' | sed '1d;s/.*\///;/^(/d')"
```
Comment (krader1961) [inline share/completions/killall.fish]: Shouldn't that be `ps u`? The `a` option lists all processes with a tty, including things like shells owned by root.

**Comment by pemrmaeoyve** [comment]
I've made the changes you suggested, but in the end, if you don't think this should be merged with master, feel free to close the pull request.

**Comment by faho** [comment]
Again - `ps u` is unsuitable, because it doesn't print processes without a tty. This includes X11 clients - I can't get firefox with this, even though it's clearly killable by my user, not just root. I'd rather have a few too many options than too few.

Looking further into this, it's _really_ complicated getting just "the name" of a process because "comm" gets truncated and "cmdline" can be altered by the process (usually it should contain the argv delimited with \0, but plenty of processes - avahi and chromium included write to it and separate with spaces instead. This makes reliably finding just argv[0] impossible - I think htop also fails here, because it can't always highlight the "basename" correctly).

What this needs is for someone to figure out what _exactly_ programs like killall expect - can I pass it "avahi-daemon: helper"? This then needs to be recreated in a manner that ideally:

- Shows no processes the user can't kill

- Shows all processes the user can kill

- Shows an argument for every process that makes killall kill that process

And that will likely be a whole lot more work.

So, in conclusion, I have to say: Sorry, but I will not be merging this.
